### PR TITLE
GODRIVER-3092 Fix UNIX socket in URL parsing.

### PR DESCRIPTION
**Pull request by qingyang-hu**
GODRIVER-3092

## Summary

- Fix bug in parsing "mongodb://%2Ftmp%2Fmongodb-27017.sock/" by discarding the `url.Parse()` call.
- Add test case
- Refactor `ConnString`

## Background & Motivation
Add `RawHosts` to `ConnString`, so we can initial `Topology` with scheme and hosts parsed by `ConnString` from the URI.
The `RawHosts` is directly parsed from the URI string before SRV lookup.

### CODE REVIEW

File: x/mongo/driver/connstring/connstring.go [hunk 1]
```diff
@@ -897,110 +586,110 @@ func (p *parser) addOption(pair string) error {
...
+			tlsssl = false
 		default:
 			return fmt.Errorf("invalid value for %q: %q", key, value)
 		}
-		if p.tlsssl != nil && *p.tlsssl != p.SSL {
+		if !u.SSLSet {
+			u.SSL = tlsssl
+		} else if tlsssl != u.SSL {
 			return errors.New("tls and ssl options, when both specified, must be equivalent")
 		}
```
Comment (qingyang-hu) [inline x/mongo/driver/connstring/connstring.go:604]: Move the detection logic here, which originally uses `*tlsssl` in `parser{}`.

File: bson/encoder_example_test.go [hunk 2]
```diff
@@ -248,7 +248,7 @@ func ExampleEncoder_IntMinSize() {
 		panic(err)
 	}
 
-	enc, err := bson.NewEncoder(vw)
+	enc := bson.NewEncoder(vw)
```
Comment (qingyang-hu) [inline bson/encoder_example_test.go:251]: To fix the build.

File: mongo/options/clientoptions.go [hunk 3]
```diff
@@ -237,7 +237,6 @@ type ClientOptions struct {
 	ZstdLevel                *int
 
 	err error
-	uri string
 	cs  *connstring.ConnString
```
Comment (qingyang-hu) [inline mongo/options/clientoptions.go:240]: Remove the `uri` so the `ConnString.Original` is the single point of truth.

**Review by matthewdale — APPROVED** [review]
Looks good! 👍

**Review by prestonvasquez — APPROVED** [review]

**Pull request merged**